### PR TITLE
cudaPackages.cutensor: 1.7.0 -> 2.0.2.4

### PR DESCRIPTION
**Pull request by cfhammill**
## Description of changes

Adds an additional new version of cutensor to cudaPackages. This was done by copying the manifest files from the previous version and modifying the paths and the hash for the x86_64-linux version.

TODO:
- fix md5 and size for the x86_64-linux version
- correct md5, size, and hash for all other versions

CC @SomeoneSerge @samuela - is there an easy way to update the redistrib json file to get the sha/md5/size for the other architectures?

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by cfhammill** [comment]
wrote a little python script to automate getting the hashes and sizes, seems to still work for x86_64-linux. The script is

```
import json
import subprocess
import copy
import os
import sys

data = json.load(sys.stdin)

dl_prefix = "https://developer.download.nvidia.com/compute/cutensor/redist/"

def get_info(path):
    o = subprocess.run(["nix-prefetch-url", "--print-path", path],
                           capture_output=True, check=True, text=True)
    sha256, store_path = o.stdout.strip().split("\n")
    fsize = os.path.getsize(store_path)

    o = subprocess.run(["nix-hash", "--type", "md5", "--base16", "--flat", store_path],
                       capture_output=True, check=True, text=True)

    md5 = o.stdout.strip()

    subprocess.run(["nix-store", "--delete", store_path], stdout=subprocess.DEVNULL)

    return {
        "sha256" : sha256,
        "md5"    : md5,
        "size"   : str(fsize)
    }


libs = set(data.keys()).difference({"release_date"})
for lib in libs:
    archs = set(data[lib].keys()).difference({"name", "license", "version"})
    for arch in archs:
        arch_info = data[lib][arch]
        path = dl_prefix + arch_info["relative_path"]
        new_info = get_info(path)
        arch_info.update(new_info)

print(json.dumps(data, indent=2))
```

happy to contrib this if something doesn't already exist and this would be useful to the cudaPackages maintainers.

**Comment by cfhammill** [comment]
Check failure seems spurious. Not sure how to trigger a rerun.

File: pkgs/development/cuda-modules/cutensor/manifests/redistrib_2.0.2.json
Comment (SomeoneSerge) [inline pkgs/development/cuda-modules/cutensor/manifests/redistrib_2.0.2.json:1]: Hmm so https://developer.download.nvidia.com/compute/cutensor/redist/redistrib_2.0.2.json shows `"release_date":	"2024-06-24"` and many other differences. Basically, we want the `redist*json` files to match upstream exactly (at the time of the merge, if nvidia chooses to update the manifests later)

Comment (SomeoneSerge) [reply pkgs/development/cuda-modules/cutensor/manifests/redistrib_2.0.2.json:1]: > correct md5, size, and hash for all other versions

We don't (unless by accident) use md5, but are you saying that the upstream manifests report incorrect sizes and hashes?

Comment (cfhammill) [reply pkgs/development/cuda-modules/cutensor/manifests/redistrib_2.0.2.json:1]: nope, just replacing the ones that I copy pasta'd from 1.7*

Comment (cfhammill) [reply pkgs/development/cuda-modules/cutensor/manifests/redistrib_2.0.2.json:1]: irrelevant now anyway, I retrieved the json from nvidia

**Comment by cfhammill** [comment]
no problem, have the correct json from nvidia now, added trailing newline to satisfy editorconfig. Should be good now.

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 324674 --extra-nixpkgs-config '{
  cudaSupport = true;
  cudaCapabilities = [
    "7.0"
    "8.0"
    "8.6"
  ];
  cudaEnableForwardCompat = false;
  allowUnfreePredicate =
    p:
    builtins.all (
      license:
      license.free
      || builtins.elem license.shortName [
        "CUDA EULA"
        "cuDNN EULA"
        "cuTENSOR EULA"
        "NVidia OptiX EULA"
      ]
    ) (if builtins.isList p.meta.license then p.meta.license else [ p.meta.license ]);
  permittedInsecurePackages = [
    "electron-24.8.6"
    "zotero-6.0.27"
  ];
}'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages failed to build:</summary>
  <ul>
    <li>cudaPackages_10.cutensor</li>
    <li>cudaPackages_10.cutensor.dev</li>
    <li>cudaPackages_10.cutensor.lib</li>
    <li>cudaPackages_10.cutensor.static</li>
    <li>python311Packages.chainer</li>
    <li>python311Packages.chainer.dist</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.cupy.dist</li>
    <li>python311Packages.vllm</li>
    <li>python311Packages.vllm.dist</li>
    <li>python312Packages.chainer</li>
    <li>python312Packages.chainer.dist</li>
    <li>python312Packages.cupy</li>
    <li>python312Packages.cupy.dist</li>
    <li>python312Packages.vllm</li>
    <li>python312Packages.vllm.dist</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.cutensor.dev</li>
    <li>cudaPackages.cutensor.lib</li>
    <li>cudaPackages.cutensor.static</li>
    <li>cudaPackages.cutensor_1_7</li>
    <li>cudaPackages.cutensor_1_7.dev</li>
    <li>cudaPackages.cutensor_1_7.lib</li>
    <li>cudaPackages.cutensor_1_7.static</li>
    <li>cudaPackages_10.cutensor_1_7</li>
    <li>cudaPackages_10.cutensor_1_7.dev</li>
    <li>cudaPackages_10.cutensor_1_7.lib</li>
    <li>cudaPackages_10.cutensor_1_7.static</li>
    <li>cudaPackages_11.cutensor</li>
    <li>cudaPackages_11.cutensor.dev</li>
    <li>cudaPackages_11.cutensor.lib</li>
    <li>cudaPackages_11.cutensor.static</li>
    <li>cudaPackages_11.cutensor_1_7</li>
    <li>cudaPackages_11.cutensor_1_7.dev</li>
    <li>cudaPackages_11.cutensor_1_7.lib</li>
    <li>cudaPackages_11.cutensor_1_7.static</li>
  </ul>
</details>

**Comment by cfhammill** [comment]
hmm, ok, so I think we need to disable this for cuda < 11, the other builds are to be expected, cupy is a dep but broken anyway, cupy being broken means vllm and chainer can't build. So the only unexpected fails are for the old cuda.

File: pkgs/development/cuda-modules/cutensor/extension.nix [hunk 1]
```diff
@@ -96,11 +97,14 @@ let
...
-    { feature, ... }:
+    { feature, redistrib, ... }:
     (attrsets.attrByPath [
       pname
       redistArch
-    ] null feature) != null;
+    ] null feature) != null
+
+    # NOTE: This is an ad hoc hack; manifest schemas do not support version constraints yet
+    && !(lib.versionOlder cudaVersion "11.0" && lib.versionAtLeast redistrib.${pname}.version "2.0.2");
```
Comment (SomeoneSerge) [inline pkgs/development/cuda-modules/cutensor/extension.nix:107]: @ConnorBaker I don't suppose it's worth the effort trying to accommodate this generically until after we've refactored cuda to have only one evalModules call

Comment (SomeoneSerge) [reply pkgs/development/cuda-modules/cutensor/extension.nix:107]: Btw older versions of cutensor are apparently broken too: https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fk4s2k2ilxf4h5rj70jm13ip7kzbyr3pl-libcutensor-1.5.0.3.drv/log?via-job=6bd1847f-fca5-4a9a-b3d4-4fac2eeba5a0

**Pull request merged**